### DISCUSSION TITLE
How can we use dependencies in timing middleware?

### DISCUSSION THREAD

**Discussion by navyamehta** [fastapi/fastapi#8537] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from fastapi_utils.timing import add_timing_middleware
app = FastAPI()
add_timing_middleware(app, record=logger.info, prefix="app", exclude="untimed")
```

### Description

Here (https://fastapi-utils.davidmontague.xyz/user-guide/timing-middleware/), a potential timing middleware is provided which accesses route and execution time.
But, what if I have a custom dependency I have written (say called `GetCurrentUserID`) and in the timing logs I also want to include the user id? I don't seem to be able to use dependencies in middleware as previous issues have pointed out since middleware is directly managed by Starlette. But, my goal of printing something like
'User ID ABCDEFG accessed route /api/print, runtime: 10 ms' can't be a dependency either since I need to record the time before and after execution to know run-time. One possible solution is to split them up, where a dependency prints 'User ID ABCDEFG accessed route /api/print' and a timing middleware prints 'route /api/print, runtime: 10 ms', but I'd like to ideally have them in a single line. Any thoughts around this? Would really appreciate insight!

### Operating System

macOS

### FastAPI Version

0.70.0

### Python Version

Python 3.7.10

**Comment by rafsaf** [comment, marked as answer]
You can use `request.state` https://www.starlette.io/requests/

`request` is reserved parameter in FastAPI and accessible in every route.

```python
from fastapi import FastAPI, Request, Response
import time

app = FastAPI()


@app.middleware("http")
async def add_process_time_header(request: Request, call_next):
    start_time = time.time()
    print(request.state.__dict__)
    # {'_state': {}}

    response: Response = await call_next(request)
    process_time = time.time() - start_time

    print(request.state.__dict__)
    # {'_state': {'user_id': 20}}

    return response


@app.get("/")
async def root(request: Request):
    request.state.user_id = 20
    return {"message": "Hello World"}


```

**Comment by navyamehta** [comment]
Thank you! Would you know any way that the dependency can be auto attached to the request? I currently have >20 endpoints and thus manually setting request.state seems very error-prone. Would really appreciate any insight!

**Comment by rafsaf** [comment]
Sorry, it was obvious for me ;)

You probably already have dependency (you mentioned it in first post), I would add the code there.

```python
from fastapi import Depends
...

def test_dep(request: Request):
    request.state.user_id = 120
    print("I'm in test_dependency and I also have access to request param as in route in prev example")

@app.get("/")
async def root(test_dep=Depends(test_dep)):
    ...
```

**Comment by navyamehta** [comment]
Ahh makes sense, so a dependency attached on the main app that injects the user_id into the request state. Thanks!

**Comment by EdgyEdgemond** [comment]
To inject additional params into the timing middleware, you'll need to customise the timing middleware. I wanted to add a common id across different timing logs.

I modified the add timing middleware to add a trace id and modified the TimingStats object to store the trace id and use it in the output logs.

```
def add_timing_middleware(
    app: FastAPI, record: Optional[Callable[[str], None]] = None,
) -> None:
    """
    Adds a middleware to the provided `app` that records timing metrics using the provided `record` callable.
    Typically `record` would be something like `logger.info` for a `logging.Logger` instance.
    """
    metric_namer = _MetricNamer(app=app)

    @app.middleware("http")
    async def timing_middleware(request: Request, call_next: RequestResponseEndpoint) -> Response:
        metric_name = metric_namer(request.scope)
        trace_id = str(uuid4())
        request.state.trace_id = trace_id
        with _TimingStats(metric_name, record=record, trace_id=trace_id):
            response = await call_next(request)
        return response
```

``` 
class _TimingStats:
    def __init__(
            self, name: Optional[str] = None, record: Callable[[str], None] = None, trace_id: Optional[str] = None,
    ) -> None:
        self.name = name
        self.record = record or print
        self.trace_id = trace_id
        
   ...
   
    def emit(self, note: Optional[str] = None) -> None:
        """
        Emit timing information, optionally including a specified note
        """
        if not self.silent:
            self.take_split()
            cpu_ms = 1000 * self.cpu_time
            wall_ms = 1000 * self.time

            message = "TIMING: Request {trace_id}" if self.trace_id is not None else "TIMING:"
            message += " | Wall: {wall_ms:6.1f}ms | CPU: {cpu_ms:6.1f}ms | {name}"

            if note is not None:
                message += " ({note})"

            self.record(
                message.format(
                    trace_id=self.trace_id,
                    wall_ms=wall_ms,
                    cpu_ms=cpu_ms,
                    name=self.name,
                    note=note,
                ),
            )
```

**Comment by EdgyEdgemond** [comment]
For clarity, I use `request.state.trace_id` elsewhere in the request lifetime to allow linking other logs to the timing logs.

**Comment by EdgyEdgemond** [comment]
This wouldn't work so well for modifying it with request based data. If you added the timer onto the request.state (as is done in the fastapi_utils.timing middleware). You could then inject the user_id into the timer once you hit your dependency.

```
    @app.middleware("http")
    async def timing_middleware(request: Request, call_next: RequestResponseEndpoint) -> Response:
        metric_name = metric_namer(request.scope)
        with _TimingStats(metric_name, record=record, exclude=exclude) as timer:
            setattr(request.state, TIMER_ATTRIBUTE, timer)
            response = await call_next(request)
        return response
```
```
def inject_user_id(request: Request) -> None:
    timer = getattr(request.state, TIMER_ATTRIBUTE, None)
    if timer is not None:
        timer.user_id = user_id
```
